arelle/plugin/CacheBuilder.py

```diff
@@ -0,0 +1,103 @@
+        return False
+    VISITED_FILEPATHS.add(filepath)
+    if filepath.startswith(cacheDirectory):
+        relativeFilepath = filepath.replace(cacheDirectory + os.sep, '')
+        destFilepath = os.path.join(outputDirectory, relativeFilepath)
+        destFilepath = os.path.abspath(destFilepath)
+        if not os.path.exists(destFilepath):
+            os.makedirs(os.path.dirname(destFilepath), exist_ok=True)
+            shutil.copyfile(filepath, destFilepath)
+    return True
+
+
+def _populateCache(modelDocument: ModelDocument, cacheDirectory: str, outputDirectory: str) -> None:
+    """
+    Recursively navigates the DTS stemming from `modelDocument` to copy referenced files into cache
+    :param modelDocument: Document to copy and search for references for recursive copying
+    :param cacheDirectory: Directory path of Arelle's cache
+    :param outputDirectory: Directory to copy cache files to
+    """
+    if _copyFileToCache(modelDocument.filepath, cacheDirectory, outputDirectory):
+        for referencedDocument, ref in modelDocument.referencesDocument.items():
+            _populateCache(referencedDocument, cacheDirectory, outputDirectory)
+
+
+def _writeCacheToZip():
+    """
+    Writes contents of cache builder output directory to `cache.zip`.
+    This does NOT include the cache output directory folder in the relative ZIP paths,
+    i.e. /http and /https directories are at the root level.
+    """
+    with zipfile.ZipFile(OUTPUT_ZIP_PATH, 'w', zipfile.ZIP_DEFLATED) as zip:
+        for root, dirs, files in os.walk(OUTPUT_DIRECTORY):
+            for file in files:
+                relativePath = os.path.relpath(os.path.join(root, file), OUTPUT_DIRECTORY)
+                zip.write(os.path.join(root, file), relativePath)
+
+
+class CacheBuilderPlugin(PluginHooks):
+    @staticmethod
+    def fileSourceFile(cntlr: Cntlr, filepath: str, *args: Any, **kwargs: Any) -> None:
+        _copyFileToCache(filepath, cntlr.webCache.cacheDir, OUTPUT_DIRECTORY)
+
+    @staticmethod
+    def validateFinally(val: ValidateXbrl, *args: Any, **kwargs: Any) -> None:
+        modelXbrl = val.modelXbrl
+        _populateCache(modelXbrl.modelDocument, modelXbrl.fileSource.cntlr.webCache.cacheDir, OUTPUT_DIRECTORY)
+
+
+# Arelle does not currently have a plugin hook that fires upon exit from all contexts: GUI, CLI, web server, etc.
+# In place of that, we can use `atexit` to trigger the writing of a ZIP file. This ensures that everything copied
+# into the cache builder output directory during the course of any Arelle application lifecycle is included in the ZIP.
+atexit.register(_writeCacheToZip)
+
+
+__pluginInfo__ = {
+    "name": PLUGIN_NAME,
+    "version": "0.0.1",
+    "description": "Generate cache archives from web documents referenced during an Arelle application lifecycle.",
+    "license": "Apache-2",
+    "author": authorLabel,
+    "copyright": copyrightLabel,
+    "Validate.Finally": CacheBuilderPlugin.validateFinally,
```

Comment:
Why is the validation hook required?

Reply:
This is the initial hook I was using before the `FileSource.File`, but I suppose it's not necessary anymore.